File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/VectorUtil.java
import java.util.*;
import java.lang.Math;

public class VectorUtil {

    public static <N> double similarity(double[] vector1, double[] vector2) {
        double[] normVec1 = normalize(vector1);
        double[] normVec2 = normalize(vector2);
        double dotProduct = 0.0;
        for (int i = 0; i < vector1.length; i++) {
            dotProduct += normVec1[i] * normVec2[i];
        }
        return dotProduct;
    }

    public static <N> List<Double> cosineSimilarities(double[] vector1, List<double[]> vectorsAll) {
        List<Double> similarities = new ArrayList<>();
        double normVec1 = norm(vector1);

        for (double[] vec : vectorsAll) {
            double normVecAll = norm(vec);
            if (normVecAll == 0.0) {
                similarities.add(0.0);
                continue;
            }
            double dotProduct = 0.0;
            for (int i = 0; i < vector1.length; i++) {
                dotProduct += vec[i] * vector1[i];
            }
            double similarity = dotProduct / (normVec1 * normVecAll);
            similarities.add(similarity);
        }
        return similarities;
    }

    public static <N> double nSimilarity(List<double[]> vectorList1, List<double[]> vectorList2) {
        if (vectorList1.isEmpty() || vectorList2.isEmpty()) {
            throw new IllegalArgumentException("At least one of the lists is empty.");
        }

        double[] meanVec1 = new double[vectorList1.get(0).length];
        double[] meanVec2 = new double[vectorList2.get(0).length];

        for (double[] vec : vectorList1) {
            for (int i = 0; i < vec.length; i++) {
                meanVec1[i] += vec[i];
            }
        }
        for (double[] vec : vectorList2) {
            for (int i = 0; i < vec.length; i++) {
                meanVec2[i] += vec[i];
            }
        }

        for (int i = 0; i < meanVec1.length; i++) {
            meanVec1[i] /= vectorList1.size();
            meanVec2[i] /= vectorList2.size();
        }

        return similarity(meanVec1, meanVec2);
    }

    public static Map<String, Double> computeIdfWeightDict(int totalNum, Map<String, Double> numberDict) {
        Map<String, Double> result = new HashMap<>();
        Map<Integer, String> index2KeyMap = new HashMap<>();
        List<Double> countList = new ArrayList<>();
        int index = 0;

        for (Map.Entry<String, Double> entry : numberDict.entrySet()) {
            index2KeyMap.put(index, entry.getKey());
            countList.add(entry.getValue());
            index++;
        }

        double[] a = new double[countList.size()];
        for (int i = 0; i < countList.size(); i++) {
            a[i] = Math.log((totalNum + 1.0) / (countList.get(i) + 1.0));
        }

        for (int i = 0; i < a.length; i++) {
            result.put(index2KeyMap.get(i), a[i]);
        }

        return result;
    }

    private static double norm(double[] vec) {
        double sum = 0.0;
        for (double val : vec) {
            sum += val * val;
        }
        return Math.sqrt(sum);
    }

    private static double[] normalize(double[] vec) {
        double[] normalized = new double[vec.length];
        double vecNorm = norm(vec);
        if (vecNorm == 0.0) {
            return normalized;
        }
        for (int i = 0; i < vec.length; i++) {
            normalized[i] = vec[i] / vecNorm;
        }
        return normalized;
    }

    public static void main(String[] args) {
        // Example usage
        double[] vector1 = {1.0, 2.0};
        double[] vector2 = {2.0, 3.0};
        System.out.println("Similarity: " + similarity(vector1, vector2));

        List<double[]> vectors = Arrays.asList(new double[]{1.0, 2.0}, new double[]{2.0, 3.0});
        System.out.println("Cosine Similarities: " + cosineSimilarities(vector1, vectors));

        List<double[]> vectorList1 = Arrays.asList(new double[]{1.0, 2.0}, new double[]{2.0, 3.0});
        List<double[]> vectorList2 = Arrays.asList(new double[]{2.0, 3.0}, new double[]{3.0, 4.0});
        System.out.println("N Similarity: " + nSimilarity(vectorList1, vectorList2));

        Map<String, Double> numberDict = new HashMap<>();
        numberDict.put("word1", 2.0);
        numberDict.put("word2", 3.0);
        System.out.println("IDF Weight Dict: " + computeIdfWeightDict(10, numberDict));
    }
}

File: ClassEval_T/cpp/solution/VectorUtil.cpp

#include <cmath>
#include <stdexcept>

#include <array>
#include <vector>
#include <unordered_map>
#include <string>

class VectorUtil {
public:
    template<size_t N>
    static double similarity(const std::array<double, N>& vector_1, const std::array<double, N>& vector_2);

    template<size_t N>
    static std::vector<double> cosine_similarities(const std::array<double, N>& vector_1,
        const std::vector<std::array<double, N>>& vectors_all);

    template<size_t N>
    static double n_similarity(const std::vector<std::array<double, N>>& vector_list_1,
        const std::vector<std::array<double, N>>& vector_list_2);

    static std::unordered_map<std::string, double> compute_idf_weight_dict(int total_num,
        const std::unordered_map<std::string, double>& number_dict);

private:
    template<size_t N>
    static double norm(const std::array<double, N>& vec);

    template<size_t N>
    static std::array<double, N> normalize(const std::array<double, N>& vec);
};


template<size_t N>
double VectorUtil::norm(const std::array<double, N>& vec) {
    double sum = 0.0;
    for (double val : vec) {
        sum += val * val;
    }
    return std::sqrt(sum);
}


template<size_t N>
std::array<double, N> VectorUtil::normalize(const std::array<double, N>& vec) {
    std::array<double, N> normalized = {};
    double vec_norm = norm(vec);
    if (vec_norm == 0.0) {
        return normalized; 
    }
    for (size_t i = 0; i < N; ++i) {
        normalized[i] = vec[i] / vec_norm;
    }
    return normalized;
}


template<size_t N>
double VectorUtil::similarity(const std::array<double, N>& vector_1, const std::array<double, N>& vector_2) {
    auto norm_vec1 = normalize(vector_1);
    auto norm_vec2 = normalize(vector_2);
    double dot_product = 0.0;
    for (size_t i = 0; i < N; ++i) {
        dot_product += norm_vec1[i] * norm_vec2[i];
    }
    return dot_product;
}


template<size_t N>
std::vector<double> VectorUtil::cosine_similarities(const std::array<double, N>& vector_1,
    const std::vector<std::array<double, N>>& vectors_all) {
    std::vector<double> similarities;
    double norm_vec1 = norm(vector_1);

    for (const auto& vec : vectors_all) {
        double norm_vec_all = norm(vec);
        if (norm_vec_all == 0.0) {
            similarities.push_back(0.0); 
            continue;
        }
        double dot_product = 0.0;
        for (size_t i = 0; i < N; ++i) {
            dot_product += vec[i] * vector_1[i];
        }
        double similarity = dot_product / (norm_vec1 * norm_vec_all);
        similarities.push_back(similarity);
    }
    return similarities;
}


template<size_t N>
double VectorUtil::n_similarity(const std::vector<std::array<double, N>>& vector_list_1,
    const std::vector<std::array<double, N>>& vector_list_2) {
    if (vector_list_1.empty() || vector_list_2.empty()) {
        throw std::invalid_argument("At least one of the lists is empty.");
    }

    std::array<double, N> mean_vec1 = {};
    std::array<double, N> mean_vec2 = {};

    for (const auto& vec : vector_list_1) {
        for (size_t i = 0; i < N; ++i) {
            mean_vec1[i] += vec[i];
        }
    }
    for (const auto& vec : vector_list_2) {
        for (size_t i = 0; i < N; ++i) {
            mean_vec2[i] += vec[i];
        }
    }

    for (size_t i = 0; i < N; ++i) {
        mean_vec1[i] /= vector_list_1.size();
        mean_vec2[i] /= vector_list_2.size();
    }

    return similarity(mean_vec1, mean_vec2);
}


std::unordered_map<std::string, double> VectorUtil::compute_idf_weight_dict(int total_num,
    const std::unordered_map<std::string, double>& number_dict) {
    std::unordered_map<std::string, double> result;
    std::unordered_map<int, std::string> index_2_key_map;
    std::vector<double> count_list;
    int index = 0;

    for (const auto& [key, count] : number_dict) {
        index_2_key_map[index] = key;
        count_list.push_back(count);
        ++index;
    }

    std::vector<double> a(count_list.size());
    for (size_t i = 0; i < count_list.size(); ++i) {
        a[i] = std::log((total_num + 1.0) / (count_list[i] + 1.0));
    }

    for (size_t i = 0; i < a.size(); ++i) {
        result[index_2_key_map[i]] = a[i];
    }

    return result;
}


template double VectorUtil::similarity(const std::array<double, 2>& vector_1, const std::array<double, 2>& vector_2);
template double VectorUtil::similarity(const std::array<double, 8>& vector_1, const std::array<double, 8>& vector_2);
template std::vector<double> VectorUtil::cosine_similarities(const std::array<double, 2>& vector_1,
    const std::vector<std::array<double, 2>>& vectors_all);
template std::vector<double> VectorUtil::cosine_similarities(const std::array<double, 8>& vector_1,
    const std::vector<std::array<double, 8>>& vectors_all);
template double VectorUtil::n_similarity(const std::vector<std::array<double, 2>>& vector_list_1,
    const std::vector<std::array<double, 2>>& vector_list_2);
template double VectorUtil::n_similarity(const std::vector<std::array<double, 8>>& vector_list_1,
    const std::vector<std::array<double, 8>>& vector_list_2);
